scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -75,110 +79,112 @@ private object KryoRegistrarLoader {
-
-        k.forSubclass[SpecificRecordBase](new SpecificAvroSerializer)
-        k.forSubclass[GenericRecord](new GenericAvroSerializer)
-        k.forSubclass[Message](new ProtobufSerializer)
-
-        k.forSubclass[LocalDate](new JodaLocalDateSerializer)
-        k.forSubclass[LocalTime](new JodaLocalTimeSerializer)
-        k.forSubclass[LocalDateTime](new JodaLocalDateTimeSerializer)
-        k.forSubclass[DateTime](new JodaDateTimeSerializer)
-
-        k.forSubclass[Path](new JPathSerializer)
-        k.forSubclass[ByteString](new ByteStringSerializer)
+  private val kryoPool: EmptyOnSerializeAtomicReference[KryoPool] =
+    new EmptyOnSerializeAtomicReference[KryoPool]() {
+      override def create(): KryoPool = {
+        val factory = new KryoFactory {
+          override def create(): Kryo = {
+            val k = KryoSerializer.registered.newKryo()
+            k.setReferences(options.referenceTracking)
+            k.setRegistrationRequired(options.registrationRequired)
+            k.forClass(new CoderSerializer(InstantCoder.of()))
+            k.forClass(new CoderSerializer(TableRowJsonCoder.of()))
+            // Java Iterable/Collection are missing proper equality check, use custom CBF as a
+            // workaround
+            k.register(classOf[Wrappers.JIterableWrapper[_]],
+                       new JTraversableSerializer[Any, Iterable[Any]]()(
+                         new JIterableWrapperCBF[Any]))
+            k.register(classOf[Wrappers.JCollectionWrapper[_]],
+                       new JTraversableSerializer[Any, Iterable[Any]]()(
+                         new JCollectionWrapperCBF[Any]))
+            // Wrapped Java collections may have immutable implementations, i.e. Guava, treat them
+            // as regular Scala collections as a workaround
+            k.register(classOf[Wrappers.JListWrapper[_]],
+                       new JTraversableSerializer[Any, mutable.Buffer[Any]])
+            k.forSubclass[SpecificRecordBase](new SpecificAvroSerializer)
+            k.forSubclass[GenericRecord](new GenericAvroSerializer)
+            k.forSubclass[Message](new ProtobufSerializer)
+            k.forSubclass[LocalDate](new JodaLocalDateSerializer)
+            k.forSubclass[LocalTime](new JodaLocalTimeSerializer)
+            k.forSubclass[LocalDateTime](new JodaLocalDateTimeSerializer)
+            k.forSubclass[DateTime](new JodaDateTimeSerializer)
+            k.forSubclass[Path](new JPathSerializer)
+            k.forSubclass[ByteString](new ByteStringSerializer)
+            k.forClass(new KVSerializer)
+            // TODO:
+            // TimestampedValueCoder
+            new AlgebirdRegistrar()(k)
+            KryoRegistrarLoader.load(k)
+
+            k
+          }
+        }
 
-        k.forClass(new KVSerializer)
-        // TODO:
-        // TimestampedValueCoder
+        new KryoPool.Builder(factory).softReferences().build()
```

Comment:
I wonder if the pool should exists in the object not the class, we must run some benchmarks and get some traces on the objects here before we merge this.

Reply:
@ravwojdyla since it's dependent on `KryoOptions` I guess for now it can be under `KryoAtomicCoder` (no strong feelings in moving it to `KryoAtomicCoder` object)